### ISSUE TITLE
Link between UKSI and National Recording Scheme species lists

### ISSUE THREAD

**Issue by DavidRoy** [BiologicalRecordsCentre/iRecord#876]
links to other UKSI tasks, e.g. #636 #409 #645 
To support reporting and filtering (e.g. verifier setup), requirement to have a 'pre-built' link between the UKSI and taxon-lists for National Recording Schemes.

Step 1.
@StephRorke to define SQL to identify taxa_taxon_list_ids for each national recording schemes.  Example sql to define UK Ladybird Scheme list (a simple one!) is
_select id, taxon, authority
from cache_taxa_taxon_lists
where taxon_list_id = 15
and family_taxon = 'Coccinellidae'_

Step 2.
@johnvanbreda to advise how to integrate within the Indicia warehouse database.  e.g. would this become a new set of taxon_groups.  Do we need to extend cache_taxa_taxon_lists to incorporate a tag for National Recording Schemes?

Step 3.
@johnvanbreda @burkmarr to build into the process for periodic update of UKSI

**Comment by johnvanbreda** [comment]
Can I confirm @DavidRoy - is the requirement here simply to make it easy to quickly add a list of taxa to a verifiers filter, or is it also considered important to build a catalogue of the recording schemes so we can use it for multiple purposes? 

In terms of the data model, the `groups` table allows us to define any collection of people for any purpose, with a filter defining the topic of interest. This is the same part of the data model as Activities. So adding "Recording scheme" to the group types termlist and providing a clone of the Activity setup forms would provide an easy way to capture the information. This has added benefits - e.g. ways of controlling membership, or country boundaries etc could all be added easily if we decide they are useful in future. The only change to the code and data model would be around caching this information to make it usable for filtering and selectable for association with each verifier's filter. 

If it's just a way of quickly adding taxa to verifier filters, we could possibly use scratchpad lists to define the taxa then provide a feature for adding a scratchpad list to the verifier's filter.

**Comment by DavidRoy** [comment]
@johnvanbreda since recording schemes are central to everything BRC does, this is a wider requirement than just verifier setup.

Is the groups table the way to go?

**Comment by johnvanbreda** [comment]
@DavidRoy in which case, yes, using groups with a "recording scheme" group type will give us a lot of flexibility. Is the existing filter builder a good enough UI for adding the associated higher taxa, or do we need something more like the scratchpad tool? I guess it depends how many higher taxa might be needed to define the list for a recording scheme. We don't need to list all the species, just the higher taxa which contain the list of species.

**Comment by DavidRoy** [comment]
@StephRorke to confirm but I think the current filter builder is sufficient to add in relevant higher taxa (Genus upwards) to each scheme.

@johnvanbreda I should have thought of this sooner, but also note that some taxa are in more than one recording scheme.  One examples is Water beetles is a scheme that spans a few other beetle schemes.  Is that problematic?

**Comment by StephRorke** [comment]
@DavidRoy , @johnvanbreda I have currently generated scheme species lists with a combination of higher taxa and species to include but am also utilising an exclude higher taxa and species option for some groups. I can look at switching to an include only approach so we can use the current filter functionality but I won't be able to get back to you immediately.
@johnvanbreda Is it possible for you to use the generated species list to populate the groups for each scheme? Although this approach might mean more work going forward when updating the UKSI taxon list.

**Comment by kitenetter** [comment]
I think this is just repeating what has been said already: it seems to me that it would be very useful for a number of reasons if our species dictionary table allowed us to look up a species, and see which recording schemes have an interest in that species. Some species will be linked to multiple schemes, others will be linked to none.

Once such a link is established it needs to be as easy as possible to update, either due to taxonomic changes or to changes in schemes.

This could then be made available in iRecord as an additional filter choice (adding a list of schemes to the "What" set of filters) for both setting up verifiers and for general use when filtering the Explore pages.

I think the groups table approach suggested above would enable this - is that correct?

The current filter builder can in theory cope with the more complex schemes, but even at family level it is onerous to repeatedly set up filters for some schemes. The scratchpad approach would be better from this point of view, but if we can provide the list of schemes as a new filter option that would be better, as then the links only needs to be updated in one place.

I may be misunderstanding the SQL approach that David suggests at the beginnign of this thread and the approach that is being taken by @StephRorke, but it seems to me that the initial SQL approach to linking taxa to schemes needs to work with TVKs rather than taxon_list_ids as we want to be able to incorporate all relevant taxa regardless of the species dictionary used for data entry.

**Comment by StephRorke** [comment]
@johnvtanbreda the majority of the scheme lists can be generated using an include only approach if we are to use the filter interface to build the groups, how is it best to provide the higher taxa lists to you, or would you prefer taxon_ids / NBNTVKs? Will we be able to associate one species with multiple schemes if required?

**Comment by johnvanbreda** [comment]
@StephRorke taxa_taxon_list table IDs would be easier for me - the system will map these to NBN TVKs for the actually queries anyway. Yes, it's OK for a species to exist in multiple schemes. This is a medium to large sized task though as it requires new code so will need someone to let me know where it sits in priorities.

**Comment by StephRorke** [comment]
@johnvanbreda  I have emailed example files to check my output meets your requirements.

**Comment by StephRorke** [comment]
@johnvanbreda we'd like to get an update on whether or not the species list format I have provided is suitable for you to work with, along with an idea of the time required by you to create the code if possible please? Thank you

**Comment by johnvanbreda** [comment]
@StephRorke I've had a look at the files you supplied - here are a couple of thoughts:
1. If a higher taxon is enough to define the list (i.e. the scheme species of interest includes everything within that higher taxon) then just specify the higher taxon, not all the children. Of course if a genus includes species that are both included and excluded then it will be necessary to specify the exact species list, but otherwise just a list of higher taxa will do.
2. No need to include synonyms, just the preferred names.

We need to decide if this link should really be "prebuilt" and stored in the data, or whether the list gets dynamically applied at reporting time. Possibly the latter?

Here are a few technical notes:
The list of taxa can be kept in a scratchpad list. 
If we then update the filter tool to support selection of the scratchpad list as a method of choosing a list of taxa, then the species list can be linked to a group (representing the recording scheme) or individuals verification filters, which would then be able to add extra criteria such as the polygon they verify in. 
We need an import tool for the scratchpad list - either just paste names when creating them, or enhance the tool to allow IDs to be pasted.
The filter code on the warehouse and the Elasticsearch proxy will both need updating to accept a scratchpad list ID.

I think this is less than 2 days work, if the primary goal is to be able to create verification filters (and other reporting filters) simply by selecting the scheme's list. When it can be done will depend on where this fits in with other priorities (@DavidRoy?).

**Comment by kitenetter** [comment]
For info, I have been using the existing scratchpad tool for creating verification filters for schemes that need a list of families or other higher taxa, such as macro- and micro-moths, and Orthoptera and allies:
https://www.brc.ac.uk/irecord/scratchpads/admin-edit

This seems to work very well, although still requires manual editing if the taxonomy changes within the UKSI (e.g. the lists of moth families have changed over time). For those schemes that are defined by a list of higher taxa the scratchpad approach has been much quicker and easier than having to repeatedly set up filters based on choosing individual families.

**Comment by johnvanbreda** [comment]
@kitenetter that's good to see. But, how do you allocate these scratchpads to verifiers so they are able to use them from the verification tool?

**Comment by kitenetter** [comment]
The scratchpad lists are available as one of the "What" options when settng up new verification filters, via:
https://www.brc.ac.uk/irecord/filters/admin-edit

This seemed to be a side-effect of some work that Andy did for CEDaR?

![image](https://user-images.githubusercontent.com/16039111/101403539-4ad73100-38cd-11eb-8c24-fc213b8cb2ed.png)

**Comment by johnvanbreda** [comment]
Ah, looks like most of what I was suggesting is already done. Other than importing Steph's lists into scratchpads, what are the additional requirements here?

**Comment by kitenetter** [comment]
@johnvanbreda I must admit I'm a bit unclear about this myself. Earlier in this discussion (e.g. https://github.com/BiologicalRecordsCentre/iRecord/issues/876#issuecomment-651012027) we suggested that the requirement was wider than just iRecord verification, but I have lost track of what the additional requirements are.

Within iRecord, if we were to set up scratchpad taxa lists for all schemes, there would then be some development needed to expose those options to all users, e.g. on the Explore pages. My understanding is that some scratchpad lists are suitable for public use and others aren't (e.g. the National Trust customised lists).

Can @StephRorke or @DavidRoy remind us what other requirements there may be?

The only additional one I can think of at the moment is to be able to define "all taxa that are not covered by a scheme". That would be useful for the purposes of export to NBN Atlas. It is not a trivial thing, especially among those taxon groups that have a mix of marine and terrestrial or freshwater taxa, where (for instance) we are already sending the 'marine' part of a taxon group to NBN but are not sending the 'freshwater' part.

**Comment by DavidRoy** [comment]
Preference for this to be built into the UKSI at source.
DavidRoy to discuss with NHM
@StephRorke can you send me the files you sent to John for this

**Comment by DavidRoy** [comment]
@ChrisRaper looping you into this thread.  We would like to make it easier to query data (i.e. within iRecord ) by taxa within each national recording schemes.  It would make most sense for this to be stored in the UKSI 'at source'.  Do you think that is possible.
@StephRorke has identified the UKSI node to define the set of taxa taxon list items for each scheme

**Comment by ChrisRaper** [comment]
Hi David, yes that's an interesting idea ... so we'd have a table of schemes and then a table of ORGANISM nodes which those schemes control? Combined with the LINEAGE field this could make your querying much easier.

**Comment by DavidRoy** [comment]
@ChrisRaper yes, that's the idea.  it would be a one to many relationship as some taxa are covered by more than one scheme - e.g. water beetles covers families also covered by other schemes.  Would you be interested in implementing this?

**Comment by ChrisRaper** [comment]
@DavidRoy - I think it would be good to host it in the UKSI, if Steph or anyone at the BRC could be tasked with maintaining it and letting me know any changes etc. We'll have to agree table layouts etc initially but let's have that conversation :)

**Comment by DavidRoy** [comment]
@ChrisRaper agreed :-)
@StephRorke - can you provide Chris with the details of the link between schemes and source UKSI node

**Comment by johnvanbreda** [comment]
Presumably this needs a many to many relationship @DavidRoy, i.e. 2 schemes can both share 1 family, and 2 families can link to 1 scheme? Or would the ability to link to a higher taxon mean that although a family might exist in 2 schemes, this would occur at different levels in the taxon hierarchy so we can get away with 1 to many?

**Comment by DavidRoy** [comment]
yes, you are right.  My mistake

**Comment by burkmarr** [comment]
I've been reviewing this thread to try to pick out the requirements. Here's what I've boiled it down to. (Please comment if you think anything is missing.)

**Data model requirements** 

- It must be possible to query the warehouse and identify all taxa covered by an RSS.
- It must be possible to query the warehouse and identify all RSS which cover a taxon.
- It must be possible to query the warehouse to identify all taxa not covered by an RSS.

**Required use cases** 

- Support reporting & filtering so that taxa can be identified by selecting an RSS, e.g. on explore and verification pages. 
- Support general custom report building (e.g.#1435)

@ChrisRaper has provided a couple of tables in the UKSI extract he sends us to support these requirements.

![image](https://user-images.githubusercontent.com/9275178/233630338-f3cfceda-fc5f-4145-a047-b1113e297d0b.png)

![image](https://user-images.githubusercontent.com/9275178/233630357-ca8d0ed4-d1ff-4805-b830-bb143f7cb8f0.png)

**Comment by johnvanbreda** [comment]
Requirements to add to your list:
* Support a many-to-many relationship between organisms and schemes so that one organism can exist in several schemes.
* Allow organisms to be identified at higher taxon level with all children then automatically included.

Additional use cases:
* Display the schemes(s) on the record details or species details pages.
* Display the relevant scheme(s) after submitting a record.

To make the display of a scheme more useful, it would be nice if we could add a URL to the scheme website, description and logo (perhaps just a URL) to the table of schemes.

**Comment by DavidRoy** [comment]
I would also favour a scheme logo being stored in the system so we can add these (when they exist) to the species details page.  There could also be other uses - e.g. linked to a verifier's name or option to include in notification etc

**Comment by burkmarr** [comment]
@johnvanbreda @robin-hutchinson - your insights into this problem would be gratefully received: given a higher order UKSI organism_key of unspecified (and variable) taxonomic rank, I want to retrieve all the UKSI descendent taxa at all ranks. Can that be done with a single SQL query?

**Comment by burkmarr** [comment]
I came up with this - which seems reasonably fast. Does it look okay to you? (In this example organism_key is for spiders.)
```
WITH RECURSIVE lower_rank AS (
	SELECT
		id,
		parent_id,
		taxon,
		taxon_rank
	FROM
		cache_taxa_taxon_lists
	WHERE
		organism_key = 'NBNORG0000070963'
	AND taxon_list_id = 15
	AND preferred = true
	UNION
		SELECT
			cttl.id,
			cttl.parent_id,
			cttl.taxon,
			cttl.taxon_rank
		FROM
			cache_taxa_taxon_lists cttl
		INNER JOIN lower_rank lr ON lr.id = cttl.parent_id 
		AND cttl.taxon_list_id = 15
		AND cttl.preferred = true
) SELECT
	*
FROM
	lower_rank;
```

**Comment by burkmarr** [comment]
Ah ha! Thanks @johnvanbreda.

**Comment by burkmarr** [comment]
@johnvanbreda can you cast your eye over this list of steps?

- Update the warehouse indicia_setup module in order to add the `scheme` and `scheme_taxa` tables in a new DB version. Fields `scheme_key` & `scheme_name` from `scheme` table and `scheme_key` & `organism_key` from `scheme_taxa` table plus the standard Indicia internal maintenance fields. May also need to create a cache table (or update an existing one) - see below.
- Add scripts to the `support_files/UKSI` scripts to populate/update the tables with data from the UKSI snapshot Access DB (and possibly update cache tables if involved).
- Update the `support_files/UKSI` data to include data files from the `scheme` and `scheme_taxa` UKSI snapshot tables.
- Update the relevant Indicia _read the docs_ to include SQL for extracting the `scheme` and `scheme_taxa` data files from the UKSI snapshot tables. 
- Once Warhouse1 database updated, run a one-off data upload in order to avoid running the whole UKSI update scripts again?
- Update the incremental UKSI update workflow/scripts.
- Consider whether enriching the Indicia taxa cache tables is worthwhile for performance and, if so, update the cache_builder module and incorporate cache table updates into UKSI full & incremental update scripts.

**Comment by ChrisRaper** [comment]
Hi David

Yes we actually did this last winter, I think, or maybe before then. The UKSI export I send in a major update includes the SCHEME and SCHEME_TAXA tables which Steph set up and I imported. The only thing we/I haven't done is to update the Import script that changes the underlying tables so that it identifies when it is doing something that will delete an ORGANISM node that is being used by a SCHEME_TAXA row.

Now we have the ORGANISM_MASTER.LINEAGE field we can SELECT entire branches from any ORGANISM_MASTER row. 🙂

Best wishes,
Chris R.
________________________________

**Comment by johnvanbreda** [comment]
@burkmarr - a few thoughts:
* Name the tables recording_scheme and recording_scheme_taxa (scheme a bit ambiguous?).
* fields id, title and description in recording_scheme for consistency (not scheme_key, scheme_name etc).
* Yes to updating the full UKSI import to handle the new tables.
* Suggest documenting how to do just the new scheme table updates from UKSI to cover your 5th bullet, so we can do it again in future.
* Not sure that scheme taxa updates will be covered by UKSI History. The only incremental update operation that would affect the scheme tables is a merge taxa operation - names are never removed, only deprecated, so incremental updates won't otherwise affect scheme taxa. By using organism key we are avoiding needing to make updates when a preferred taxon name change occurs.
* As described in emails, I'd suggest not adding scheme information to the cache tables but adding a filter to the standard parameters system which pre-selects the list of taxon meaning IDs for a scheme's organism keys, then uses the taxon meaning IDs in a taxon paths filter. For the equivalent elasticsearch filter we'd currently need to map the organism keys to external keys (accepted name TVKs) and filter on that in the same way the other taxonomy filters work (in ElasticsearchProxyHelper.php). In future we ought to consider putting the organism key into Elasticsearch documents so we can use that as a more stable way of filtering.

**Comment by burkmarr** [comment]
@johnvanbreda @DavidRoy - the email correspondence and summary by John in the previous comment make good sense to me. I propose that I tackle the steps to create and populate `recording_scheme` and `recording_scheme_taxa` warehouse tables first. And when that's done to our satisfaction, we can close this issue. Then we can raise new issues to make use of the data, e.g. updating the standard parameters system and elasticsearch filters. @johnvanbreda - shall I branch from dev to do the warehouse work?

**Comment by burkmarr** [comment]
@johnvanbreda - I merged this code to create two new RSS tables into the develop branch in June. It was originally for version 8.20, but you later changed that to 8.21 because a hotfix had claimed 8.20. The Warehouse is now on version 8.22.3 but when I checked, I found that the RSS tables had not been created. The folder `modules/indicia_setup/db/version_8_21_0` is not in the master branch so presumably develop has not been merged into master and the current version results from a further hotfeature or hotfix. Do you have any feeling for when the develop branch will be merged into master?

**Comment by johnvanbreda** [comment]
Am hoping to do a couple more iRecord issues, then do a full merge later this week. If you need it more quickly let me know as it's only a loose plan.

**Comment by burkmarr** [comment]
@johnvanbreda - no that's perfect. Don't worry if you have to delay either.

**Comment by burkmarr** [comment]
With the recent release of the new warehouse, the new `recording_schemes` and `recording_scheme_taxa` are now present in the database and I have run the script to populate them with the data from UKSI Access snapshot. So the warehouse now has the information we need to relate taxa to RSS.

@DavidRoy - I suggest that this issue is closed now and new issues raised to make use of the data (like #1435 which I hope to work on this week).

**Comment by DavidRoy** [comment]
Thanks Rich

**Comment by burkmarr** [comment]
@johnvanbreda - I've reopened this issue because there are some loose ends. When I created the report for https://github.com/BiologicalRecordsCentre/iRecord/issues/1435 to retrieve the RSS for a given taxon_meaning_id, the report initially failed because there was no read permission for `indicia_report_user` on the `recording_schemes` and `recording_scheme_taxa` tables. To fix it I manually granted these permissions using pgAdmin, but I guess that isn't the way it should be done.

I looked at other DB SQL scripts that create tables in the warehouse `indicia_setup` module, e.g. indicia_setup/db/version_8_16_0/202302161350_local_rules.sql, but I can't see anything in there which grants the permissions that exist on the tables they create. So where should this step be performed?
